feat: Improve error messages in connector execution #311

jvallesm opened the pull request:
Console errorrs are unclear and expose implementation details to no-code / low-code users.

This PR proposes a way to bubble up errors from components in a way that the end-user error is separated from the internal error information (stack, low-level details). The `sendReq` method in `pkg/openai` is modified to return human-friendly errors. https://github.com/instill-ai/pipeline-backend/pull/311 is needed to test these changes.

~The [fault](https://github.com/Southclaws/fault) package is chosen to carry out this task.~ While `fault` was chosen originally, some of the features we were using were no longer needed after refactoring the code to adapt to the use of `temporal` workflows to execute sync pipelines.

- A logger is injected in the `openai.Client` struct, so we no longer need to pass the request variables as part of the error context.
- The client doesn't know the component ID and can't attach it to the end-user message. However, because communication between temporal processes prevents error unwrapping, we can only use temporal errors to pass the end-user message.

Therefore, it was simpler to use [our own implementation](https://github.com/instill-ai/x/pull/13) of error messages, at least while the required functionality doesn't get too sofisticated.

### 🗒️ Notes

- Some coverage was added due to the rapid development in the `main` branches, which made manual testing a bit unreliable. Let me know if this is a potential change preventer and if we should focus on coverage later.
  - Additionally, the `frankban/quicktest` package was added in the tests. I'm a fan of the package but I can rework the tests with `require` in order to align with the current tests.
- Many integration tests were removed as they were outdated. They aren't being run now, if there's a reason to keep them I can revert that change.

![CleanShot 2023-12-15 at 14 45 30](https://github.com/instill-ai/pipeline-backend/assets/3977183/da89cd77-b4b8-401e-aae8-506c55ef0290)


![CleanShot 2023-12-15 at 14 44 32](https://github.com/instill-ai/pipeline-backend/assets/3977183/5512bf25-6bd6-4f64-abdd-d61ee3f35e33)

jvallesm committed:
chore: Extract HTTPClient interface to util

donch1989 commented:
Hi @jvallesm 
1. Please feel free to remove any outdated tests, and we can go for `frankban/quicktest`.
2. I'm thinking we can put the shared error package in [x](https://github.com/instill-ai/x). This is a repo that is shared by all of our repos. The [component](https://github.com/instill-ai/x) is actually only used in pipeline-backend.
3. There are some problems in the current test cases. Most of the connectors need the vendors' `api_key` or `token` to be testable. It will be great if we can change it to load those keys from environment variable, and we can easily inject it via CI or other method in the future.

jvallesm commented:
Hi @donch1989, thanks for the insights 🙇 

> There are some problems in the current test cases. Most of the connectors need the vendors' api_key or token to be testable. It will be great if we can change it to load those keys from environment variable, and we can easily inject it via CI or other method in the future

Yes that's something I realised, in the integration tests API keys were being loaded from `test_artifacts/`, it seemed a bit odd to load them this way instead of through environment variables. For the current tests a dummy key was enough as I'm only testing the 401 response but I agree later we'll need to use real API keys (or fake the APIs and test against the real ones only on certain pipelines).

jvallesm committed:
feat: extract OpenAI request details from error to log

The request details aren't usually relevant for end users, so instead of
exposing them as part of the error we log them before returning a
simpler and conciser error.

jvallesm committed:
feat: add human-friendly message to OpenAI non-2xx responses

The [errmsg](https://github.com/instill-ai/x/errmsg) package is used to
improve the clarity of errors caused by non-2xx HTTP responses. This
improves the UX and hides the implementation details of the connector.

`pipeline-backend` will then search for end-user messages within the
error when handling errors from connectors.

This commit removes the integration tests, which didn't compile nor
succeed, and adds an integration test for OpenAI API errors. New
integration tests can be added as functionality changes. The logic
implemented before this commit can be referenced for new integration
tests.

jvallesm commented on pkg/integration_test.go:
```diff
@@ -4,468 +4,63 @@
-	op, err := hfCon.Execute([]*structpb.Struct{&in})
-	fmt.Printf("\n op :%v, err:%s", op, err)
-}
+	exec, err := conn.CreateExecution(uid, "TASK_TEXT_GENERATION", config, logger)
```
❓ Does it make sense to pass the logger when `conn` already has an exported logger? I don't think we'll ever have a use case where we'll want different loggers.

I kept the signature because it's aligned with the other `CreateExecution` statements.

donch1989 replied:
I think we can remove all the `logger` argument from `CreateExecution`, we can setup the logger defined [here](https://github.com/instill-ai/component/blob/c139af45d2a7f6e4117e8ec8ac355cac207630b5/pkg/base/component.go#L75) when `Init()` and reuse it

jvallesm replied:
👍 This requires modifying [the `IComponent` interface](https://github.com/instill-ai/component/blob/main/pkg/base/component.go#L33). As a refactorisation task that touches several repos, I'd go for a separate task (along with the comment below). I updated the linear ticket to reflect this.

donch1989 replied:
No problem

jvallesm commented on pkg/openai/main.go:
```diff
@@ -87,15 +85,22 @@ func (c *Connector) CreateExecution(defUID uuid.UUID, task string, config *struc
 
 // NewClient initializes a new OpenAI client
-func NewClient(apiKey, org string) Client {
+func NewClient(apiKey, org string, logger *zap.Logger) Client {
```
`logger` is passed here as an argument, as it's the strategy in this repo. In e.g. `pipeline-backend` we use `zap.GetLogger`, which would be possible here. Would it make sense to unify the strategies?

donch1989 replied:
Yes, we need to propagate the logger from pipeline-backend to here.

jvallesm replied:
👍 There's [a ticket](https://linear.app/instill-ai/issue/INS-2960/propagate-logger-to-connectors-potentially-to-any-component) for this, it can be tackled separately or as the starting point of the next error handling PR (my plan was validating and merging this, then working a new PR with one commit per connector)

donch1989 replied:
Great!

jvallesm commented on pkg/openai/main.go:
```diff
@@ -104,22 +109,53 @@ func (c *Client) sendReq(reqURL, method, contentType string, data io.Reader) ([]
-		err = fmt.Errorf("non-200 status code: %d, while calling URL: %s, response body: %s", res.StatusCode, reqURL, respBody)
-		return nil, err
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
```
ℹ️ I used a broader range of statuses as there's no contract around OpenAI response codes.

donch1989 reviewed (approved):
LGTM
